### PR TITLE
Issue #9396: Add example of AST for TokenTypes.AT

### LINKED ISSUE #9396: Add example of AST for TokenTypes.AT
https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#AT

needs to be extended to show example of AST.
Source to be updated: https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java

PLEASE READ for more details: https://github.com/checkstyle/checkstyle/wiki/Good-practice-on-how-to-update-AST-in-TokenTypes

### PR DESCRIPTION
**Pull request by vinayak-gaikwad**
Resolves #9396
```
vinayak@Gladiator:~/Documents$ cat Test.java
class Test{
	public static void main(String args[]){
		Derived d1 = new Derived();
		d1.set(1);
	}
}

class Base {
	protected int num;
    public void set() {}
}

class Derived extends Base {
    @Override
    public void set(int n) {
        this.num = n;
    }
}


vinayak@Gladiator:~/Documents/Opensource/checkstyle/target$ java -jar checkstyle-8.45-SNAPSHOT-all.jar -T /home/vinayak/Documents/Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|--LITERAL_CLASS -> class [1:0]
|--IDENT -> Test [1:6]
`--OBJBLOCK -> OBJBLOCK [1:10]
    |--LCURLY -> { [1:10]
    |--METHOD_DEF -> METHOD_DEF [2:1]
    |   |--MODIFIERS -> MODIFIERS [2:1]
    |   |   |--LITERAL_PUBLIC -> public [2:1]
    |   |   `--LITERAL_STATIC -> static [2:8]
    |   |--TYPE -> TYPE [2:15]
    |   |   `--LITERAL_VOID -> void [2:15]
    |   |--IDENT -> main [2:20]
    |   |--LPAREN -> ( [2:24]
    |   |--PARAMETERS -> PARAMETERS [2:25]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [2:25]
    |   |       |--MODIFIERS -> MODIFIERS [2:25]
    |   |       |--TYPE -> TYPE [2:25]
    |   |       |   |--IDENT -> String [2:25]
    |   |       |   `--ARRAY_DECLARATOR -> [ [2:36]
    |   |       |       `--RBRACK -> ] [2:37]
    |   |       `--IDENT -> args [2:32]
    |   |--RPAREN -> ) [2:38]
    |   `--SLIST -> { [2:39]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [3:2]
    |       |   |--MODIFIERS -> MODIFIERS [3:2]
    |       |   |--TYPE -> TYPE [3:2]
    |       |   |   `--IDENT -> Derived [3:2]
    |       |   |--IDENT -> d1 [3:10]
    |       |   `--ASSIGN -> = [3:13]
    |       |       `--EXPR -> EXPR [3:15]
    |       |           `--LITERAL_NEW -> new [3:15]
    |       |               |--IDENT -> Derived [3:19]
    |       |               |--LPAREN -> ( [3:26]
    |       |               |--ELIST -> ELIST [3:27]
    |       |               `--RPAREN -> ) [3:27]
    |       |--SEMI -> ; [3:28]
    |       |--EXPR -> EXPR [4:8]
    |       |   `--METHOD_CALL -> ( [4:8]
    |       |       |--DOT -> . [4:4]
    |       |       |   |--IDENT -> d1 [4:2]
    |       |       |   `--IDENT -> set [4:5]
    |       |       |--ELIST -> ELIST [4:9]
    |       |       |   `--EXPR -> EXPR [4:9]
    |       |       |       `--NUM_INT -> 1 [4:9]
    |       |       `--RPAREN -> ) [4:10]
    |       |--SEMI -> ; [4:11]
    |       `--RCURLY -> } [5:1]
    `--RCURLY -> } [6:0]
CLASS_DEF -> CLASS_DEF [8:0]
|--MODIFIERS -> MODIFIERS [8:0]
|--LITERAL_CLASS -> class [8:0]
|--IDENT -> Base [8:6]
`--OBJBLOCK -> OBJBLOCK [8:11]
    |--LCURLY -> { [8:11]
    |--VARIABLE_DEF -> VARIABLE_DEF [9:1]
    |   |--MODIFIERS -> MODIFIERS [9:1]
    |   |   `--LITERAL_PROTECTED -> protected [9:1]
    |   |--TYPE -> TYPE [9:11]
    |   |   `--LITERAL_INT -> int [9:11]
    |   |--IDENT -> num [9:15]
    |   `--SEMI -> ; [9:18]
    |--METHOD_DEF -> METHOD_DEF [10:4]
    |   |--MODIFIERS -> MODIFIERS [10:4]
    |   |   `--LITERAL_PUBLIC -> public [10:4]
    |   |--TYPE -> TYPE [10:11]
    |   |   `--LITERAL_VOID -> void [10:11]
    |   |--IDENT -> set [10:16]
    |   |--LPAREN -> ( [10:19]
    |   |--PARAMETERS -> PARAMETERS [10:20]
    |   |--RPAREN -> ) [10:20]
    |   `--SLIST -> { [10:22]
    |       `--RCURLY -> } [10:23]
    `--RCURLY -> } [11:0]
CLASS_DEF -> CLASS_DEF [13:0]
|--MODIFIERS -> MODIFIERS [13:0]
|--LITERAL_CLASS -> class [13:0]
|--IDENT -> Derived [13:6]
|--EXTENDS_CLAUSE -> extends [13:14]
|   `--IDENT -> Base [13:22]
`--OBJBLOCK -> OBJBLOCK [13:27]
    |--LCURLY -> { [13:27]
    |--METHOD_DEF -> METHOD_DEF [14:4]
    |   |--MODIFIERS -> MODIFIERS [14:4]
    |   |   |--ANNOTATION -> ANNOTATION [14:4]
    |   |   |   |--AT -> @ [14:4]
    |   |   |   `--IDENT -> Override [14:5]
    |   |   `--LITERAL_PUBLIC -> public [15:4]
    |   |--TYPE -> TYPE [15:11]
    |   |   `--LITERAL_VOID -> void [15:11]
    |   |--IDENT -> set [15:16]
    |   |--LPAREN -> ( [15:19]
    |   |--PARAMETERS -> PARAMETERS [15:20]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [15:20]
    |   |       |--MODIFIERS -> MODIFIERS [15:20]
    |   |       |--TYPE -> TYPE [15:20]
    |   |       |   `--LITERAL_INT -> int [15:20]
    |   |       `--IDENT -> n [15:24]
    |   |--RPAREN -> ) [15:25]
    |   `--SLIST -> { [15:27]
    |       |--EXPR -> EXPR [16:17]
    |       |   `--ASSIGN -> = [16:17]
    |       |       |--DOT -> . [16:12]
    |       |       |   |--LITERAL_THIS -> this [16:8]
    |       |       |   `--IDENT -> num [16:13]
    |       |       `--IDENT -> n [16:19]
    |       |--SEMI -> ; [16:20]
    |       `--RCURLY -> } [17:4]
    `--RCURLY -> } [18:0]

```
Screenshot from 2021-07-19 13-22-24

### CODE REVIEW

**Comment by romani** [comment]
Please put in javadoc a bit more nodes, to show a context 
```
--MODIFIERS -> MODIFIERS [14:4]
    |   |   |--ANNOTATION -> ANNOTATION [14:4]
    |   |   |   |--AT -> @ [14:4]
    |   |   |   `--IDENT -> Override [14:5]
    |   |   `--LITERAL_PUBLIC -> public [15:4]
```

**Comment by vinayak-gaikwad** [comment]
Resolves #9396

```
vinayak@Gladiator:~/Documents$ cat Test.java
class Test{
	public static void main(String args[]){
		Derived d1 = new Derived();
		d1.set(1);
	}
}

class Base {
	protected int num;
    public void set() {}
}

class Derived extends Base {
    @Override
    public void set(int n) {
        this.num = n;
    }
}


vinayak@Gladiator:~/Documents/Opensource/checkstyle/target$ java -jar checkstyle-8.45-SNAPSHOT-all.jar -T /home/vinayak/Documents/Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|--LITERAL_CLASS -> class [1:0]
|--IDENT -> Test [1:6]
`--OBJBLOCK -> OBJBLOCK [1:10]
    |--LCURLY -> { [1:10]
    |--METHOD_DEF -> METHOD_DEF [2:1]
    |   |--MODIFIERS -> MODIFIERS [2:1]
    |   |   |--LITERAL_PUBLIC -> public [2:1]
    |   |   `--LITERAL_STATIC -> static [2:8]
    |   |--TYPE -> TYPE [2:15]
    |   |   `--LITERAL_VOID -> void [2:15]
    |   |--IDENT -> main [2:20]
    |   |--LPAREN -> ( [2:24]
    |   |--PARAMETERS -> PARAMETERS [2:25]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [2:25]
    |   |       |--MODIFIERS -> MODIFIERS [2:25]
    |   |       |--TYPE -> TYPE [2:25]
    |   |       |   |--IDENT -> String [2:25]
    |   |       |   `--ARRAY_DECLARATOR -> [ [2:36]
    |   |       |       `--RBRACK -> ] [2:37]
    |   |       `--IDENT -> args [2:32]
    |   |--RPAREN -> ) [2:38]
    |   `--SLIST -> { [2:39]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [3:2]
    |       |   |--MODIFIERS -> MODIFIERS [3:2]
    |       |   |--TYPE -> TYPE [3:2]
    |       |   |   `--IDENT -> Derived [3:2]
    |       |   |--IDENT -> d1 [3:10]
    |       |   `--ASSIGN -> = [3:13]
    |       |       `--EXPR -> EXPR [3:15]
    |       |           `--LITERAL_NEW -> new [3:15]
    |       |               |--IDENT -> Derived [3:19]
    |       |               |--LPAREN -> ( [3:26]
    |       |               |--ELIST -> ELIST [3:27]
    |       |               `--RPAREN -> ) [3:27]
    |       |--SEMI -> ; [3:28]
    |       |--EXPR -> EXPR [4:8]
    |       |   `--METHOD_CALL -> ( [4:8]
    |       |       |--DOT -> . [4:4]
    |       |       |   |--IDENT -> d1 [4:2]
    |       |       |   `--IDENT -> set [4:5]
    |       |       |--ELIST -> ELIST [4:9]
    |       |       |   `--EXPR -> EXPR [4:9]
    |       |       |       `--NUM_INT -> 1 [4:9]
    |       |       `--RPAREN -> ) [4:10]
    |       |--SEMI -> ; [4:11]
    |       `--RCURLY -> } [5:1]
    `--RCURLY -> } [6:0]
CLASS_DEF -> CLASS_DEF [8:0]
|--MODIFIERS -> MODIFIERS [8:0]
|--LITERAL_CLASS -> class [8:0]
|--IDENT -> Base [8:6]
`--OBJBLOCK -> OBJBLOCK [8:11]
    |--LCURLY -> { [8:11]
    |--VARIABLE_DEF -> VARIABLE_DEF [9:1]
    |   |--MODIFIERS -> MODIFIERS [9:1]
    |   |   `--LITERAL_PROTECTED -> protected [9:1]
    |   |--TYPE -> TYPE [9:11]
    |   |   `--LITERAL_INT -> int [9:11]
    |   |--IDENT -> num [9:15]
    |   `--SEMI -> ; [9:18]
    |--METHOD_DEF -> METHOD_DEF [10:4]
    |   |--MODIFIERS -> MODIFIERS [10:4]
    |   |   `--LITERAL_PUBLIC -> public [10:4]
    |   |--TYPE -> TYPE [10:11]
    |   |   `--LITERAL_VOID -> void [10:11]
    |   |--IDENT -> set [10:16]
    |   |--LPAREN -> ( [10:19]
    |   |--PARAMETERS -> PARAMETERS [10:20]
    |   |--RPAREN -> ) [10:20]
    |   `--SLIST -> { [10:22]
    |       `--RCURLY -> } [10:23]
    `--RCURLY -> } [11:0]
CLASS_DEF -> CLASS_DEF [13:0]
|--MODIFIERS -> MODIFIERS [13:0]
|--LITERAL_CLASS -> class [13:0]
|--IDENT -> Derived [13:6]
|--EXTENDS_CLAUSE -> extends [13:14]
|   `--IDENT -> Base [13:22]
`--OBJBLOCK -> OBJBLOCK [13:27]
    |--LCURLY -> { [13:27]
    |--METHOD_DEF -> METHOD_DEF [14:4]
    |   |--MODIFIERS -> MODIFIERS [14:4]
    |   |   |--ANNOTATION -> ANNOTATION [14:4]
    |   |   |   |--AT -> @ [14:4]
    |   |   |   `--IDENT -> Override [14:5]
    |   |   `--LITERAL_PUBLIC -> public [15:4]
    |   |--TYPE -> TYPE [15:11]
    |   |   `--LITERAL_VOID -> void [15:11]
    |   |--IDENT -> set [15:16]
    |   |--LPAREN -> ( [15:19]
    |   |--PARAMETERS -> PARAMETERS [15:20]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [15:20]
    |   |       |--MODIFIERS -> MODIFIERS [15:20]
    |   |       |--TYPE -> TYPE [15:20]
    |   |       |   `--LITERAL_INT -> int [15:20]
    |   |       `--IDENT -> n [15:24]
    |   |--RPAREN -> ) [15:25]
    |   `--SLIST -> { [15:27]
    |       |--EXPR -> EXPR [16:17]
    |       |   `--ASSIGN -> = [16:17]
    |       |       |--DOT -> . [16:12]
    |       |       |   |--LITERAL_THIS -> this [16:8]
    |       |       |   `--IDENT -> num [16:13]
    |       |       `--IDENT -> n [16:19]
    |       |--SEMI -> ; [16:20]
    |       `--RCURLY -> } [17:4]
    `--RCURLY -> } [18:0]
```
image

**Comment by romani** [comment]
Single commit in PR, and keep only nodes I referenced in my comment

**Comment by vinayak-gaikwad** [comment]
> Single commit in PR, and keep only nodes I referenced in my comment

should I close this PR and create a new one?

**Comment by romani** [comment]
Do not close PR, do fixes on your local and push with force to GitHub to the same branch.

https://checkstyle.org/beginning_development.html#Starting_Development

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 1]
```diff
@@ -4896,6 +4896,18 @@
      * to the interface literal signifying the definition of an annotation
      * declaration.
      *
+     * <p>For example:</p>
+     * <pre> @Override </pre>
+     *
+     * <p>parses as:</p>
+     * <pre>
+     * --MODIFIERS -&gt; MODIFIERS
+     *      |   |   |--ANNOTATION -&gt; ANNOTATION
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:4905]: Remove "--"

Remove "|" that goes to nowhere.

Indent all right

**Review by romani — CHANGES_REQUESTED** [review]
Item:

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 2]
```diff
@@ -4939,6 +4939,18 @@
      * to the interface literal signifying the definition of an annotation
      * declaration.
      *
+     * <p>For example:</p>
+     * <pre> @Override </pre>
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:4943]: we need to show structure in real life context, make it: 
```
@Deprecated
private int value; 
```

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 2 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Done.

**Comment by rnveach** [comment]
I rebased the branch and fixed the issue brought up.

**Review by romani — APPROVED** [review]

**Pull request merged**